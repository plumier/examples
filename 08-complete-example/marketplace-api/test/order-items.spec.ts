import supertest from "supertest"
import createApp from "../src/app"
import { closeConnection, createCart, createUser, omitKeys } from "./_helper"

const omit = (obj: any) => omitKeys(obj, ["createdAt", "updatedAt"])

afterEach(async () => {
    await closeConnection()
})

describe("Order Items", () => {

    it("Should accessible by resource owner", async () => {
        const app = await createApp({ mode: "production" })
        const { user, shops: [putra] } = await createCart(app)
        await supertest(app.callback())
            .post("/api/carts/checkout")
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body: orders } = await supertest(app.callback())
            .get(`/api/shops/${putra.shop.id}/orders`)
            .set("Authorization", `Bearer ${putra.staffs[0].token}`)
            .expect(200)
        const {body} = await supertest(app.callback())
            .get(`/api/orders/${orders[0].id}/items`)
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        expect(omit(body)).toMatchSnapshot()
    })

    it("Should accessible by shop staff", async () => {
        const app = await createApp({ mode: "production" })
        const { user, shops: [putra] } = await createCart(app)
        await supertest(app.callback())
            .post("/api/carts/checkout")
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body: orders } = await supertest(app.callback())
            .get(`/api/shops/${putra.shop.id}/orders`)
            .set("Authorization", `Bearer ${putra.staffs[0].token}`)
            .expect(200)
        await supertest(app.callback())
            .get(`/api/orders/${orders[0].id}/items`)
            .set("Authorization", `Bearer ${putra.staffs[0].token}`)
            .expect(200)
    })

    it("Should accessible by shop owner", async () => {
        const app = await createApp({ mode: "production" })
        const { user, shops: [putra] } = await createCart(app)
        await supertest(app.callback())
            .post("/api/carts/checkout")
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body: orders } = await supertest(app.callback())
            .get(`/api/shops/${putra.shop.id}/orders`)
            .set("Authorization", `Bearer ${putra.staffs[0].token}`)
            .expect(200)
        await supertest(app.callback())
            .get(`/api/orders/${orders[0].id}/items`)
            .set("Authorization", `Bearer ${putra.owner.token}`)
            .expect(200)
    })

    it("Should not accessible by other user", async () => {
        const app = await createApp({ mode: "production" })
        const { user, shops: [putra] } = await createCart(app)
        const otherUser = await createUser(app, { email: "john.doe@ymail.com" })
        await supertest(app.callback())
            .post("/api/carts/checkout")
            .set("Authorization", `Bearer ${user.token}`)
            .expect(200)
        const { body: orders } = await supertest(app.callback())
            .get(`/api/shops/${putra.shop.id}/orders`)
            .set("Authorization", `Bearer ${putra.staffs[0].token}`)
            .expect(200)
        await supertest(app.callback())
            .get(`/api/orders/${orders[0].id}/items`)
            .set("Authorization", `Bearer ${otherUser.token}`)
            .expect(403)
    })
})