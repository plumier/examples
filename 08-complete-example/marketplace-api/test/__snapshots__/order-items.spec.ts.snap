// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Items Should accessible by resource owner 1`] = `
Array [
  Object {
    "deleted": false,
    "id": 1,
    "price": 2500,
    "product": Object {
      "deleted": false,
      "id": 1,
      "name": "MacBook Pro 16 2020",
      "price": 2500,
    },
    "quantity": 1,
    "shop": Object {
      "deleted": false,
      "id": 1,
      "name": "Putra Mahkota",
    },
    "subTotal": 2500,
  },
  Object {
    "deleted": false,
    "id": 2,
    "price": 250,
    "product": Object {
      "deleted": false,
      "id": 2,
      "name": "Li-Ion Battery",
      "price": 250,
    },
    "quantity": 2,
    "shop": Object {
      "deleted": false,
      "id": 1,
      "name": "Putra Mahkota",
    },
    "subTotal": 500,
  },
  Object {
    "deleted": false,
    "id": 3,
    "price": 900,
    "product": Object {
      "deleted": false,
      "id": 3,
      "name": "M3 SSD 1TB",
      "price": 900,
    },
    "quantity": 2,
    "shop": Object {
      "deleted": false,
      "id": 1,
      "name": "Putra Mahkota",
    },
    "subTotal": 1800,
  },
]
`;
